packages/vite/src/node/plugins/css.ts

```diff
@@ -912,7 +912,9 @@ export function cssPostPlugin(config: ResolvedConfig): Plugin {
           css += chunkCSSMap.get(chunk.preliminaryFileName) ?? ''
         }
 
-        for (const chunkName of chunkCSSMap.keys())
+        // sorted chunk names to guarantee output order (deterministic output)
+        const chunkNames = [...chunkCSSMap.keys()].sort()
+        for (const chunkName of chunkNames)
```

Comment:
We cannot sort chunk names alphabetically because the order of the chunks are meaningful. The issue with `chunkCSSMap` being non-deterministic is because it's being populated via `renderChunk` that runs non-deterministically among chunks.

I think a better solution here is instead of iterating `chunkCSSMap` here, we can iterate `prelimaryNameToChunkMap` instead which is deterministic as it's derived from the `bundle..

Reply:
`prelimaryNameToChunkMap` is also not deterministic, entry points are in deterministic order, but the chunks are randomly sorted.

Reply:
Can you share more how you got that? Chunks in Rollup (regardless if entry, non-entry, dynamic entry) should be consistent, otherwise it'll break many usecases and assumptions in Vite currently.

Reply:
Hmm testing manually I can also see that `bundle` isn't deterministic. However, `renderChunk`'s chunks are and `chunkCSSMap` should be following that order and should be deterministic. Unless there's an additional Vite plugin that uses async `renderChunk()` and changes the order, which is a bug I found today.

But we still can't sort alphabetically like this change now. It will break, for example, if you import two CSS files and expect them to render in order.

To fix this, it might be worth making `chunkCSSMap` deterministic directly, but it should be fixed since we added [this line](https://github.com/vitejs/vite/blob/c3b3d3c617ba729814c599e14145499aeac4c9d3/packages/vite/src/node/plugins/css.ts#L679-L685) though. (Unless you use plugins that with the bug I mentioned above)

Reply:
I checked my custom plugins but they do not seem to change the order.

<details>
```
 plugins: [
'vite:build-metadata',
'vite:watch-package-data',
'vite:pre-alias',
'alias',
'vite:modulepreload-polyfill',
'vite:resolve',
'vite:html-inline-proxy',
'vite:css',
'vite:esbuild',
'vite:json',
'vite:wasm-helper',
'vite:worker',
'vite:asset',
'vite:vue2',
'vite-plugin-node-polyfills',
'esbuild-minify',
undefined,
'vite:wasm-fallback',
'vite:define',
'vite:css-post',
'vite:build-html',
'vite:worker-import-meta-url',
'vite:asset-import-meta-url',
'vite:force-systemjs-wrap-complete',
'commonjs',
'vite:data-uri',
'inject',
'vite:dynamic-import-vars',
'vite:import-glob',
'remove-ensure-watch-plugin',
'rollup-plugin-license',
'vite:build-import-analysis',
'vite:esbuild-transpile',
'vite:terser',
'vite:reporter',
'vite:load-fallback']
```
</details>

---

I changed the approach a bit by not sorting but extracting from entries and then use their imports.
This seems to be deterministic as the entry order seems to be stable.

Reply:
IIRC I created a test in Rollup and can verify that dynamic entries and its chunks are non-deterministic as Rollup processes the chunks in parallel. But anyways I think the current fix is also good enough, I'm not sure of a better way otherwise.